File: Sources/TaskLensCore/Events.swift
import Foundation

public enum EventType: String, Codable {
    case taskRegistered = "TASK_REGISTERED"
    case taskSubmitted = "TASK_SUBMITTED"
    case taskEnqueued = "TASK_ENQUEUED"
    case taskWaiting = "TASK_WAITING"
    case taskEligible = "TASK_ELIGIBLE"
    case taskLaunched = "TASK_LAUNCHED"
    case taskStarted = "TASK_STARTED"
    case taskStopped = "TASK_STOPPED"
    case taskExpired = "TASK_EXPIRED"
    case taskRetryRequested = "TASK_RETRY_REQUESTED"
    case taskRescheduled = "TASK_RESCHEDULED"
    case taskSucceeded = "TASK_SUCCEEDED"
    case taskFailed = "TASK_FAILED"
    case taskCancelled = "TASK_CANCELLED"
    case taskCompletionReported = "TASK_COMPLETION_REPORTED"

    case constraintChanged = "CONSTRAINT_CHANGED"
    case networkChanged = "NETWORK_CHANGED"
    case batteryChanged = "BATTERY_CHANGED"
    case powerModeChanged = "POWER_MODE_CHANGED"
    case appStateChanged = "APP_STATE_CHANGED"
    case processStateChanged = "PROCESS_STATE_CHANGED"
    case backgroundRefreshChanged = "BACKGROUND_REFRESH_CHANGED"

    case urlSessionTransferStarted = "URLSESSION_TRANSFER_STARTED"
    case urlSessionTransferCompleted = "URLSESSION_TRANSFER_COMPLETED"
    case urlSessionTransferFailed = "URLSESSION_TRANSFER_FAILED"

    case httpRequestStarted = "HTTP_REQUEST_STARTED"
    case httpRequestCompleted = "HTTP_REQUEST_COMPLETED"
    case httpRequestFailed = "HTTP_REQUEST_FAILED"

    case customBreadcrumb = "CUSTOM_BREADCRUMB"
}

public enum EventSource: String, Codable {
    case workManager = "WORK_MANAGER"
    case bgTaskScheduler = "BG_TASK_SCHEDULER"
    case jobScheduler = "JOB_SCHEDULER"
    case foregroundService = "FOREGROUND_SERVICE"
    case alarmManager = "ALARM_MANAGER"
    case urlSession = "URL_SESSION"
    case systemBroadcast = "SYSTEM_BROADCAST"
    case application = "APPLICATION"
    case kourier = "KOURIER"
    case custom = "CUSTOM"
}

public enum EventSeverity: String, Codable {
    case info = "INFO"
    case warning = "WARNING"
    case error = "ERROR"
    case critical = "CRITICAL"
}

public struct TaskLensEvent: Codable, Identifiable {
    public let id: String
    public let taskId: String?
    public let attemptId: String?
    public let timestamp: Date
    public let sequenceNumber: Int64
    public let type: EventType
    public let source: EventSource
    public let severity: EventSeverity
    public let attributes: [String: String]
    public let schemaVersion: Int

    private static var sequenceCounter: Int64 = 0
    private static let lock = NSLock()

    public static func nextSequence() -> Int64 {
        lock.lock()
        defer { lock.unlock() }
        sequenceCounter += 1
        return sequenceCounter
    }

    public static func resetSequence(to value: Int64 = 0) {
        lock.lock()
        defer { lock.unlock() }
        sequenceCounter = value
    }

    public init(
        id: String = UUID().uuidString,
        taskId: String? = nil,
        attemptId: String? = nil,
        timestamp: Date = Date(),
        sequenceNumber: Int64? = nil,
        type: EventType,
        source: EventSource,
        severity: EventSeverity = .info,
        attributes: [String: String] = [:],
        schemaVersion: Int = 1
    ) {
        self.id = id
        self.taskId = taskId
        self.attemptId = attemptId
        self.timestamp = timestamp
        if let seq = sequenceNumber {
            self.sequenceNumber = seq
        } else {
            self.sequenceNumber = Self.nextSequence()
        }
        self.type = type
        self.source = source
        self.severity = severity
        self.attributes = attributes
        self.schemaVersion = schemaVersion
    }
}

File: tasklens-core-kmp/src/commonMain/kotlin/dev/shushant/tasklens/core/Events.kt
package dev.shushant.tasklens.core

import kotlin.time.Clock
import kotlin.time.Instant
import kotlinx.atomicfu.atomic
import kotlinx.serialization.Serializable

@Serializable
enum class EventType {
    TASK_REGISTERED,
    TASK_SUBMITTED,
    TASK_ENQUEUED,
    TASK_WAITING,
    TASK_ELIGIBLE,
    TASK_LAUNCHED,
    TASK_STARTED,
    TASK_STOPPED,
    TASK_EXPIRED,
    TASK_RETRY_REQUESTED,
    TASK_RESCHEDULED,
    TASK_SUCCEEDED,
    TASK_FAILED,
    TASK_CANCELLED,
    TASK_COMPLETION_REPORTED,

    CONSTRAINT_CHANGED,
    NETWORK_CHANGED,
    BATTERY_CHANGED,
    POWER_MODE_CHANGED,
    APP_STATE_CHANGED,
    PROCESS_STATE_CHANGED,
    BACKGROUND_REFRESH_CHANGED,

    URLSESSION_TRANSFER_STARTED,
    URLSESSION_TRANSFER_COMPLETED,
    URLSESSION_TRANSFER_FAILED,

    HTTP_REQUEST_STARTED,
    HTTP_REQUEST_COMPLETED,
    HTTP_REQUEST_FAILED,

    CLOCK_SHIFT_DETECTED,
    CONFIGURATION_WARNING,
    INTERNAL_DIAGNOSTIC,

    CUSTOM_BREADCRUMB
}

@Serializable
enum class EventSource {
    WORK_MANAGER,
    BG_TASK_SCHEDULER,
    JOB_SCHEDULER,
    FOREGROUND_SERVICE,
    ALARM_MANAGER,
    URL_SESSION,
    SYSTEM_BROADCAST,
    APPLICATION,
    KOURIER,
    CUSTOM
}

@Serializable
enum class EventSeverity {
    INFO,
    WARNING,
    ERROR,
    CRITICAL
}

/** Monotonically incrementing sequence generator backed by atomicfu — safe for all KMP targets. */
private object SequenceCounter {
    private val value = atomic(0L)
    fun next(): Long = value.incrementAndGet()
    fun reset(initialValue: Long = 0L) { value.value = initialValue }
}

@Serializable
data class TaskLensEvent(
    val id: String = generateId(),
    val taskId: String? = null,
    val attemptId: String? = null,
    val timestamp: Instant = Clock.System.now(),
    val sequenceNumber: Long = SequenceCounter.next(),
    val type: EventType,
    val source: EventSource,
    val severity: EventSeverity = EventSeverity.INFO,
    val attributes: Map<String, String> = emptyMap(),
    val schemaVersion: Int = 1
) {
    companion object {
        fun nextSequence(): Long = SequenceCounter.next()
        fun resetSequence(initialValue: Long = 0L) {
            SequenceCounter.reset(initialValue)
        }
    }
}

/** Pure-Kotlin UUID v4-like ID generator using Random — no java.util.UUID dependency. */
private fun generateId(): String {
    val bytes = ByteArray(16) { (kotlin.random.Random.nextInt(256).toByte()) }
    bytes[6] = ((bytes[6].toInt() and 0x0f) or 0x40).toByte()
    bytes[8] = ((bytes[8].toInt() and 0x3f) or 0x80).toByte()
    return bytes.toHex()
}

private fun ByteArray.toHex(): String {
    val sb = StringBuilder(36)
    forEachIndexed { i, b ->
        if (i == 4 || i == 6 || i == 8 || i == 10) sb.append('-')
        sb.append(b.toInt().and(0xff).toString(16).padStart(2, '0'))
    }
    return sb.toString()
}
